man/fi_peer.3.md

```diff
@@ -0,0 +1,194 @@
+of the exporter which was returned by fi_cq_open() is provided.  On output,
+the importer will return its own cq structure.  The importer's CQ should be
+configured to report completions through the exporter.  As a result, the
+CQ functions for reading the importer CQ (e.g. fi_cq_read, fi_cq_readerr,
+etc.) should be set to return -FI_ENOSYS when called.
+
+The importer is responsible for setting up the callbacks with the exporter.
+It does this by calling fi_export_fid() to export the callbacks.  The
+importer passes in the input cq passed to fi_cq_open().  On output, the
+exporter will return a struct fid_peer_cq.  See the example flow further
+down.
+
+```c
+struct fi_ops_cq_write {
+	size_t	size;
+	ssize_t (*write)(struct fid_peer_cq *cq, void *context, uint64_t flags,
+			size_t len, void *buf, uint64_t data, uint64_t tag,
+			fi_addr_t src);
+	ssize_t	(*writeerr)(struct fid_peer_cq *cq,
+			const struct fi_cq_err_entry *err_entry);
+};
+
+struct fi_ops_cq_progress {
+	size_t	size;
+	void	(*progress)(struct fid_peer_cq *cq);
+
+};
+
+struct fid_peer_cq {
+	struct fid fid;
+	struct fi_ops_cq_write *export_ops;
+	struct fi_ops_cq_progress *import_ops;
+};
+```
+struct fi_ops_cq_write is filled out by the exporter and used by the
+importer to insert new completions into the CQ.  The exporter is
+responsible for locking, event signaling, and handling CQ overflow.
+(TBD: should CQ overflow push back to the importer for flow control).
+When the write callbacks are invoked, the data is relative to the
+importer.  For example, the fi_addr_t is relative to the importer's
+AV.  The exporter is responsible for converting the address if source
+addressing is needed.
+
+struct fi_ops_cq_progress is completed by the importer and must be
+initialized before returning from fi_cq_open().  The importer owns
+the struct fid_peer_cq::fid fields, with the exception of the size/class
+fields.
+
+## EXAMPLE PEER CQ SETUP
+
+The above description defines the generic mechanism for sharing CQs
+between providers.  This section outlines one possible implementation
+to demonstrate the use of the APIs.  In the example, the rxm provider
+uses the shm provider as a peer for data transfers targeting endpoints
+on the local node.
+
+```
+1. rxm is configured to use shm as a peer for local communication
+2. The application calls fi_cq_open(rxm_domain, &rxm_cq, ...)
+3. rxm allocates its cq and automatically configures it to be used
+   as a peer cq.
+4. rxm calls fi_cq_open(shm_domain, rxm_cq, ...), setting the
+   FI_PEER_IMPORT flag.
+5. shm allocates a cq, but configures it such that all completions
+   are written to a peer cq.  The cq ops to read from the cq are
+   set to enosys calls.
+6. shm calls fi_export_fid(rxm_cq, &peer_cq, ...) to export the peer
+   cq from the rxm provider.
+7. rxm allocates, initializes, and returns the peer_cq.  rxm creates
+   a reference between the peer_cq object and its rxm_cq object.
```

Comment:
I still think it is more intuitive to make these two changes:
(1) move 6, 7 before 4 and let rxm call fi_export_fd; 
(2) change 4 to pass peer_cq instead of rxm_cq.

Nothing else needs to change.

Reply:
I'm not following how to move 6,7 before 4.  Something must call into shm before it calls fi_export_fid.

We also need to define how the object lifetimes work.  In the model above the shm cq takes a reference on the rxm cq (done via fi_export_fid).  That reference is held until the the peer_cq is closed.  The peer_cq would be closed when closing the shm cq.

We need: Determine which provider owns the allocation of peer_cq.  Which way the object references go.  When the references are taken/released.

Reply:
Why fi_export_fid needs to be called by shm? rxm can (and should) do it.  With this change step 6 and 7 has nothing to do with shm  and can be moved before 4. 

Calling fi_export_fid from rxm also takes reference of rxm_cq. That doesn't change. The reference is held by the peer_cq. The close method of peer_cq should release the reference.

Reply:
Moving 7 before 4, pass peer_cq at 5, then remove 6 should be the same overall flow.  There would be no calls to export/import, as that is assumed by the IMPORT flag.

Reply:
Yes, but I though 7 was just explaining the internal activities of 6.

Reply:
Ah, I rememeber.  The fi_cq_open() call takes a fid_cq ** as the input.  The fid_peer_cq is a different structure.  To make the simplified flow work, the fi_cq_attr would need to be extended (most reasonable path) or use the context field.

Reply:
Whatever we come up with here should also be usable with shared rx, meaning we have this:
```
fi_srx_context(struct fid_domain *domain, struct fi_rx_attr *attr,
	       struct fid_ep **rx_ep, void *context)
```
function to work with.  We could do something like:
```
struct fi_peer_cq_attr {
    struct fi_cq_attr attr;
    struct fid_peer_cq *peer_cq;
};
```
(same with fi_rx_attr) and use the flag to know that the structure can be cast.  It would need to be more complex than this to handle forward/backward compatibility changes to the base attr structure.

Reply:
Do we need to specify `ep` to created shared rx context?

I thought rx context will be shared by multiple end points.

Reply:
@shefty Yes I like the idea of extending the attrs structures. 
@wzamazon I don't  think ep as an input is needed for creating shared rx context. shared context is associated with endpoint via fi_ep_bind().